### ISSUE TITLE
Update for modern d3

### ISSUE THREAD

**Issue by davenquinn** [davenquinn/d3-ternary#2]
Hi @JulesBlm, I figured I'd move the conversation here since it charts the future of this module. Thanks for your involvement and excitement to push the state of ternary-plotting code forward. I'd like to "do this right" in a collaborative sense, because I think having good processes for this will make the geoscience field much better at sharing progress! I've added you as a collaborator, so you should be able to directly push to the repository. Let me know if I need to adjust privileges.

I think some of the API changes you highlight in your [Observable notebook](https://observablehq.com/@julesblm/introducing-d3-ternary-plot-not-yet) make sense to move this forward. I especially like the separation of calculation and plotting code. Some of the current code internally works similarly, but it is a little more coupled than your new version.

I've laid out a suggested approach below but I am happy to her about other directions you might have in mind. Once we chart a basic course, I'm happy to initiate some of the maintenance things that should happen before we move forward, such as updating module bundling. I have a bunch of other modules like [`@macrostrat/column-components`](https://github.com/UW-Macrostrat/column-components/) that follow a rhyming structure; these will provide a good template for these [annoying](https://github.com/UW-Macrostrat/ui-components/issues/1) bits.

## Near term priorities

- [x] Tag the current version as `v1` and begin constructing `v2`
- [x] We should probably "decaffeinate" the codebase. This will require a new code bundling strategy ([Webpack](https://webpack.js.org/), [Rollup](https://rollupjs.org/guide/en/), [Parcel v2](https://v2.parceljs.org/), or equivalent), and might be a good time to add in typescript support (although this is optional). We don't have to fully utilize Typescript, but it's handy for editor support.
- [x] Make the basics work with d3 > v4. You've already done this, it will just be a matter of integrating into the module and tuning up the examples.
- [x] Basic docs — maybe this can just be a tuned-up Observable notebook!
- [x] NPM publish!

## Blue-sky/future evolution

I'm excited to see some of the more advanced functionality you mentioned (reversed axes, right-angle ternaries) and others I have had in mind (partial/composite ternaries), at some point. Keeping these in mind as we re-architect the basics will help us make sure the API can be extended to support these.

It might be handy to provide a React frontend (maybe as a separate module, or maybe within the same repository?) for the basic module. I think a nice "gold-standard" implementation of this shared-backend approach is Susie Lu's pair of [React-Annotation](https://react-annotation.susielu.com/) and [d3-Annotation](https://d3-annotation.susielu.com/), which are both based on a basic [viz-annotation](https://github.com/susielu/viz-annotation) backend.

**Comment by JulesBlm** [comment]
Great! Likewise, thanks for your enthusiasm and collaboration! 

I’ve continued the development in [another Observable notebook](https://observablehq.com/@julesblm/scratchpad-d3-ternary-plot-ii),  I’ll push to this repository once things are a bit more stable.
I previously asked Torben Jansen who made [this Observable notebook of a ternary plot](https://observablehq.com/@toja/d3-ternary-plot) for a license and his feedback. He was generous enough to take a look and suggested that the barycentric coordinates should be handled in a separate function which I think was an excellent idea, so I’ve split this too.
The function `barycentric` handles the barycentric coordinate system (converting to and inverting from barycentric coordinates), and does not handle any presentational elements (ie rendering). Calculations for rendering are performed inside `ternaryPlot()` which takes in a `barycentric` function as a parameter (hope this make sense). 

This way by updating the barycentric coordinate system, the ternary plot can be zoomed and panned. I’ve almost got zooming/panning working all right (see [this notebook by dixonj13](https://observablehq.com/@dixonj13/zoomable-ternary-plot) which formed the basis of the functionality).
). Keeping the transform in sync with the domains is something I am still figuring out, while it works, it just isn’t as clean as I’d like.

The position of the vertices of `barycentric` coord system can be changed manually as well. Try dragging the red vertices in the linked Observable notebook and watching the barycentric coordinated update (credits to [Torben Jansen](https://observablehq.com/@toja/) for this) so right-angle ternary plots should already work. 

### Bundling

D3 modules use Rollup for bundling. D3 chief Mike Bostock provides a template for d3 ‘plugins’ in [this blogpost](https://bost.ocks.org/mike/d3-plugin/). Although it’s already 5 years old I think it still a solid basis. Other 3rd party d3 modules – like [d3-geo-scalebar](https://github.com/HarryStevens/d3-geo-scale-bar) and [d3-tricontour](https://github.com/Fil/d3-tricontour – use Rollup too. It might be a good idea to ask in the d3 development slack on what's recommended. 

There's a bit on testing in the blogpost too but that's for later.
> By convention, D3 modules use Substack’s Tape as a testing harness. (And we use JSDOM for DOM testing.) You’re free to use something else, but it’s nice if modules are consistent.

## Plan

Somewhere this week I will clean up the code and make an alpha branch for the next version and tag the current branch as v1. 

Splitting it into like d3- and react- package with a common share like d3-react annotation as you mentioned is an interesting possibility. What I have so far is more akin to the ‘annotion-viz’ part, there’s surprisingly little d3 in it so far. 
The scales in ‘ternaryPlot’ are only used as a convenient way to generate ticks from a given domain and a there are few calls to `d3.line()`. I was playing around with this in React and had a basic plot working fairly quickly. This could be developed into more user-friendly component later.

### Documentation
I think examples are the best documentation, besides documentation in the readme (similar to other d3 modules), I was thinking about making several demo notebooks, including a few that make use of other d3 modules.

* Basic ternary plot with d3
* Ternary plot with zooming & panning
* Basic example custom domain
* Ternary plot with reversed domains
* Update the current examples of d3 ternary
* Ternary plot with hexagonal binning with d3-hexbin (already [done](https://observablehq.com/@julesblm/ternary-plot-with-hexagonal-binning) but needs to be updated to reflect latest changes)
* Ternary plot with contour lines with d3-contour

**Comment by davenquinn** [comment]
Hey — I'll look at some of the links you provided but a few quick thoughts now:

All the documentation examples you suggest would be fabulous. I'd like to add a "partial" ternary to the list but that might be what you had in mind for your custom example.

For bundling, I concur that Rollup is the way to go. The config [here](https://github.com/UW-Macrostrat/ui-components/blob/master/rollup.config.js) is one that I've invested some time in, so it could serve as a good and relatively up-to-date starting point. But I'm sure you will have some good additions.

**Comment by JulesBlm** [comment]
Great, I'll make a start in this repository tomorrow I’ll look into bundling then. Configuring bundlers isn't exactly my favorite task so your starter config is appreciated.

And you're right, by 'custom domain' I meant a partial ternary plot. I've just got this working nicely, even syncing nicely with d3-zoom, check it in [the updated notebook](https://observablehq.com/@julesblm/scratchpad-d3-ternary-plot-ii).

I also played around making [a right-angle ternary plot](https://observablehq.com/@julesblm/right-angle-ternary-plot) using the new module, and got it working fairly quickly.

**Comment by JulesBlm** [comment]
Ok, I've got the basics working fine, there's still a few rough edges. I have a very basic rollup setup at the moment, I'm not so sure whether dependencies should be bundled along or listed as external.

These are some issues that need to be worked out
- [x] There's a bug in the boundary condition for the transform causing the plot to shift too far right or left when it's dragged near the vertices. I based this code on [Zoomable Ternary Plot](https://observablehq.com/@dixonj13/zoomable-ternary-plot) notebook, this one doesn't have the same bug, so I'm doing something different but I can't figure out where.
- [x] Reversed domains, this can be done by swapping the vertices clockwise but this makes the transform bc invalid. I still need to think about how to best solve this.
- [ ] Make axisLabel methods more flexible to allow different locations for the labels.
- [x] Documentation: I like how Harry Stevens provides documentation for his d3-modules: [d3-regression](https://observablehq.com/@harrystevens/introducing-d3-regression) and [d3-geo-scale-bar](https://observablehq.com/@harrystevens/introducing-d3-geo-scale-bar). I think I'll make something similar.

**Comment by JulesBlm** [comment]
Been a while since I had time to work on this. I haven't worked on d3-ternary itself that much but I have figured out how to make it work with React. I've extracted the d3-ternary logic into its own hook `useTernaryPlot`, and extracted the grid, axes and ticks into separate components. I also made components for plotting of data. Here's a snippet that might make it more clear.

```javascript
function useTernaryPlot({ radius, labels, domains }) {
  const { current: ternaryProjection } = useRef(
    ternary()
      .a((d) => d.a)
      .b((d) => d.b)
      .c((d) => d.c)
  );

  const plot = useMemo(
    () =>
      ternaryPlot(ternaryProjection)
        .radius(radius)
        .gridLineCounts([20, 20, 20])
        .tickCounts([10, 10, 10])
        .labels(labels),
    [radius, labels, domains]
  );

  return { ternaryProjection, plot };
}

export function TernaryPlot({
  points,
  lines,
  areas,
  text,
  width,
  labels,
  domains,
}) {
  const radius = width / 2.1;
  const yOffset = radius / 4;

  const { ternaryProjection, plot } = useTernaryPlot({
    radius,
    labels,
    domains,
  });

  const axisLabels = plot.axisLabels();
  const gridLines = plot.gridLines();
  const ticks = plot.ticks();
  const trianglePath = useMemo(() => plot.trianglePath(), [radius]);

  const scalePoint = useCallback(
    (d) => {
      const [x, y] = ternaryProjection(d);

      return [x * radius, y * radius];
    },
    [radius]
  );

  return (
    <svg
      height={width}
      width={width}
      id="ternary-plot"
      fontFamily="sans-serif"
      className="mx-auto my-0"
    >
      <g transform={`translate(${width / 2} ${width / 1.9 + yOffset})`}>
        <clipPath id="triangle-path">
          <path d={trianglePath} />
        </clipPath>
        <AxisLabels axisLabels={axisLabels} />
        <GridLines gridLines={gridLines} />
        <AxisTicks ticks={ticks} />
        <path d={trianglePath} stroke="black" fill="none" id="triangle" />
        <g className="data" clipPath="url(#triangle-path)">
          <Areas scalePoint={scalePoint} areas={areas} />
          <Points scalePoint={scalePoint} points={points} />
          <Lines scalePoint={scalePoint} lines={lines} />
          <Text scalePoint={scalePoint} text={text} />
        </g>
      </g>
    </svg>
  );
}
```

I'll leave out the more advanced features like zooming and reversed domains for now so I can focus on polishing and documenting what I have so far. Hopefully I'll find some time for this in the coming weeks.

**Comment by JulesBlm** [comment]
Progress was slow but I managed to finish a 'beta' version that includes zooming, partial plots, and reversed domains. I developed it on Observable (see [this notebook](https://observablehq.com/@julesblm/d3-ternary-plot)) and I'll try to get to pushing it to this repository sometime soon. I wrote an [introduction notebook](https://observablehq.com/@julesblm/introducing-d3-ternary-plot) and a rough [documentation notebook](https://observablehq.com/@julesblm/ternary-plot-documentation) and [a few examples](https://observablehq.com/collection/@julesblm/ternary-plots).

**Comment by davenquinn** [comment]
Man, @JulesBlm, this is an awesome set of work. Let me know if you want some help packaging this up, doing some testing, and publishing. It could make sense to structure this as a multi-package repository, with a "core" module alongside d3 and react wrappers. But you might have a better sense of which way you want to go.

**Comment by JulesBlm** [comment]
Thanks @davenquinn! I've pushed the latest version today and added the docs to the readme. A few points 

## Code structure / quality 

### What should the function returned by ternaryPlot() do?
It took a while to wrap my head around the 'd3 way' of writing code. One thing I'm not sure of is what the function returned by `ternaryPlot()` should do. Right now it's just a configurable function, calling `ternaryPlot()` itself does't do anything. Should I move the methods that actually return the data (ie ticks, gridlines axisLabels in there). Should I move the `convert(*data*)` method in there, so it resembles `barycentric()`? Of course it's a matter of judgement/preference, I'm curious to hear what you think.

### The \`.transform()` method is buggy
This is probably where 90% of the bugs are. I want to limit the zoom to never exceed the 'initial triangle', sounds simple but it's a lot trickier than it seems. What I have now works *fine* most of the time, but occasionally some weird things happen when zooming.

### Testing
Frankly, my knowledge on testing JS is thin. I will take a look at how other d3 modules are tested and then tinker around.

## Bundling 

### Big bundle

The build seems a bit large and appears to include a lot of d3-scale and d3-array functions. I'm not sure if this is correct of if I'm using `nodeResolve()` incorrectly in rollup. Again, any help is appreciated here.

### D3 and React render functions

I've added a few reuseable 'render functions' that draw the data returned by `ternaryPlot()` the d3 way.

I've also added a React hook and React components for drawing a ternary plot. For the time being, I've just added these files and not included them in the build or added React as a dependency. I have to think about how to go about packaging this. A multi-package approach might be a good idea. It'd be great if it could be bundled like
```javascript
import { ternaryPlot, barycentric } from "d3-ternary-plot";
import { useTernaryPlot, AxisLabels, AxisTicks, GridLines, Lines  } from "@d3-ternary-plot/react"; 
// or for the d3 render functions
import { grid, axisLabels, ticks } from "@d3-ternary-plot/d3
```

But then again I'm not that familiar with bundling and publishing to npm. If you are, then I'm open to suggestions

On a final note, I propose to change the name to 'd3-ternary-plot" as it's more explicit but I'm not authorized to change the repository name. Once it's renamed and the bundling is worked out, I think it can be published to NPM.

**Comment by davenquinn** [comment]
Hi — I'll think about a few of these issues in more detail later, and I'm happy to help with packaging and testing. A few surface-level comments:

- I'm not necessarily averse to a name change, but I think `d3-ternary` is just as descriptive as `d3-ternary-plot`. And terser, which is a win for me! I guess people could get confused with the ternary operator, but I think this is unlikely. And, since a lot of the functionality is just coordinate manipulation, it's not really "just" plotting either. Both ways work, I think.
- I don't know what the default function should do, but maybe it should just be the d3 version? Or it could be the basic axis object. I'll take a look when I poke around.
- Those bundling issues don't seem insurmountable, I'll take a look to see if I can improve the rollup config.
- For packaging, one intermediate way to go would be "sub-packages" so we don't have to go through the rigamarole of creating a namespace etc. In this case `@d3-ternary/react` would become `d3-ternary/react`. The problem with this approach is we lose some of the ability to separate dependencies. So actually, namespacing is probably better.
- For namespaced packages, I think the right naming scheme would be `@d3-ternary/core`, `@d3-ternary/plot` (for d3 plot), and `@d3-ternary/react`...but i could be convinced otherwise.

**Comment by JulesBlm** [comment]
Hi Daven,

* Fair enough, `d3-ternary` is also a good name, and indeed a tad shorter. 
* Yes, such things usually become evident when playing around with it.
* I'll fiddle around with bundling some more and try to improve it.
* I am not familiar with subpackages or namespaces in npm packages. I don't think it makes a big difference, the bundles aren't that big, so let's go for whichever is easiest. I can take a look today. The d3/react annotation package you suggested might be a good example I'll take a look at that.
* The naming scheme you suggested sounds good.

**Comment by JulesBlm** [comment]
I got sidetracked by a couple of things, the two relevant sidetracks: finishing the new [TernaryPlot.com](https://ternaryplot.com/) and the other is adding TypeScript to d3-ternary. I'm not proficient in TypeScript just yet so it's taking a while.

A quick to-do list
- [x] Add TypeScript
- [ ] Improve bundling and make subpackages
- [ ] Improve React hook
- [x] NPM Publish
- [x] Small testing set-up

Then there's the overall goal of making the code less messy, I reckon that will happen with use over time. I've thought about what the default function `ternaryPlot()` should return and come to the conclusion that it should be the current `convert()` method. I think this is best because it resembles the way the `barycentric()` function and normal d3 scales work.

```javascript
const values = [33, 33, 33];

const coords = ternaryPlot(values) // = [0,0]
```

**Comment by JulesBlm** [comment]
Oh and I just managed to fix the rollup config, the trick was to list all d3 dependencies in the 'external' keyword and drop the `nodeResolve` plugin.

**Comment by davenquinn** [comment]
I think this makes sense. I've made great strides with Typescript in the year since I picked it up; I think the secret thus far is to not strive for perfection and only be rigorous about defining types at the important boundaries. I've also had some success bundling types with Rollup; I'll be happy to take a look at these aspects.

Sometime (hopefully end of this week) I will try to pull the code down and make a demo so I can test out + maybe make some improvements to those aspects in particular if needed.

I agree on the default function.

**Comment by JulesBlm** [comment]
Sounds good. I haven't yet surpassed the point of 'it's really slowing me down' with TypeScript, I'll get there. Something else I thought of to make it more d3-like is to drop the `tickCounts(*counts*)` and `gridCounts(*counts*)` and instead let the `ticks()` and `gridLines()` methods handle the counts with a paramater, similar to a d3-scale. Here the *counts* parameter can be either a number or an array of numbers.

So instead of this

```javascript
const plot = ternaryPlot(barycentric)
     .gridLineCounts(10)
     .tickCounts([15, 10, 20]);

const ticks = plot.ticks();
const gridLines = plot.gridLines();
```

It becomes this

```javascript
const plot = ternaryPlot(barycentric)

const ticks = plot.ticks([15, 10, 20]);
const gridLines = plot.gridLines(10);
```

I think this is nicer as there are already a lot of configuration methods and better resembler the [d3-scale ticks](https://github.com/d3/d3-scale/#continuous_ticks) API.

**Comment by davenquinn** [comment]
I agree with this API change, it does look significantly nicer.

**Comment by JulesBlm** [comment]
Ok, I got the TypeScript version working, there's still a few `any`'s and some other type wonkiness that I need to go over once more. Other than that, there's only testing, then I think it's ready for release.

### Release to do
- [x] Tests
- [ ] Bundling of d3 and React render functions
- [x] Publish to npm

**Comment by davenquinn** [comment]
@JulesBlm congratulations on the new release! Looks fantastic!

**Comment by JulesBlm** [comment]
Thanks a lot Daven! I'm glad it's out. Of course there's a few remaining tasks but for the most part it's 'done'. I just fiddled around with the rollup config some more and published it to npm. There's two output modules now, one d3 module (used like `d3.ternaryPlot(b)`) and another 'standalone' version that bundles the dependencies (d3-scale and d3-array) too. There might be a better way make an ES module bundle but for now this will do. And I still need to find out how to make subpackages for the render functions though.

Also the test setup is very basic, still need to expand that a little more. Feel free to give feedback or improve these when you play around with it.

**Comment by JulesBlm** [comment]
I improved the npm bundle a bit and added correct type declaration. I’m done working on this for a while and I'm closing this issue.  I'll make separate issues for the loose ends.

Feel free to go through it and make suggestions and modifications. If you’re strapped for time, as most academics are I suppose, take your time, no hurry.

I like to take this opportunity to say thank you for writing the original d3-ternary and making it open-source. Without the original d3-ternary ternaryplot.com couldn’t have existed and I likely wouldn’t have been in to JS as much as I am now.

A quick search on Google Scholar shows several papers that credit it and I’m sure there are many more that have used it. Over the years I have received messages from many students, academics and professionals expressing their gratitude. I count ~11 000 plot downloads since October 2020 so it has helped many people. That goes to show the power of open source, even for such small and niche projects that don’t draw a lot of attention!
